### ISSUE TITLE
Unexpected behavior of jax.numpy.linalg.eigh for complex Hermitian vs. real matrices

### ISSUE THREAD

**Issue by shoumikdc** [google/jax#16262]
### Description

Hello! I am using JAX for a project that involves working with Hermitian matrices. I noticed some strange behavior of the function `jax.numpy.linalg.eigh` for real matrices cast as either real or complex. I will demonstrate with the following minimum working example. 

Consider
```
a = jnp.array([[1, 1], [1, 1]], dtype=jnp.float64)
b = jnp.array([[1, 1], [1, 1]], dtype=jnp.complex128)
```

If I compute the eigenvalues of these two matrices, I get the following results:
```
jnp.linalg.eigh(a)
(Array([0., 2.], dtype=float64),
 Array([[-0.70710678,  0.70710678],
        [ 0.70710678,  0.70710678]], dtype=float64))

jnp.linalg.eigh(b)
(Array([0., 2.], dtype=float64),
 Array([[-0.70710678-0.70710678j,  0.70710678+1.20710678j],
        [ 0.70710678+0.70710678j,  0.70710678+0.5j       ]],      dtype=complex128))
```

One crucial thing to point out here is that the eigenvectors of `b` are not orthonormal! 

-----------------------------------------

I see similar buggy behavior for `jax.scipy.linalg.eigh` as well. I'll also note that if I recreate this MWE using just regular numpy arrays, the issue is present there too (using `numpy.linalg.eigh`). **However, if I use regular scipy (`scipy.linalg.eigh`), then I get identical correct results for both matrices `a` and `b`.** 

Is this a bug in how the eigenvectors are being computed? Why does `scipy.linalg.eigh` give identical correct results for both real and complex, while all of the others (numpy, jax.numpy, jax.scipy) seem to give buggy outputs for the complex case. 

### What jax/jaxlib version are you using?

jax v0.4.11, jaxlib v0.4.11

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.9.13. MacOS 12.0.1. I am using Apple Silicon.

**Comment by jakevdp** [comment]
Related issues:  #12156 #13392

I suspect this is an OSX-specific issue, and is related to whatever lapack you are using on your system. The code produces correct results in other environments, e.g. Colab CPU & GPU. I'd try updating scipy to the most recent version, since JAX uses LAPACK via scipy.

**Comment by PhilipVinc** [comment]
FYI with latest `scipy==1.10.1` and `jax==0.4.11` on an up to date Mac (13.3.1) with M2 processor the output of `jnp.linalg.eigh` is correct for both real and complex input.

**Comment by shoumikdc** [comment]
Thank you @jakevdp and @PhilipVinc! After playing around with it more last night, I also agree that it was a `scipy` issue on Apple Silicon macOS. I was able to resolve the specific problem I posted about above by re-installing `numpy`, `scipy`, and `jax` in a fresh environment. I can also confirm that the output of `jnp.linalg.eigh` gives the same results now for the problem above with `jax==0.4.11` and `scipy==1.10.1`. 

-------------------

I did have one other follow up question. Maybe I can open another thread about this, but I'll still ask here. 

Even resolving the real/complex issue above for real hermitian matrices, there is still a general problem of sign ambiguity for the returned eigenvectors. In general, the returned normalized eignvectors one gets are `[±V₁, ±V₂, ..., ±Vₙ]` (more generally one could multiply each by a complex phase, but let's assume just + or - for now). 

One thing I am noticing for slightly larger size problems is that the choice of sign/phase on each returned eigenvector differs depending on whether I specify`float64` or `complex 128` for a real hermitian matrix. In particular, I am testing both types for each of (i) `np.linalg.eigh`, (ii) `sp.linalg.eigh`, (iii) `jnp.linalg.eigh` and (iv) `jsp.linalg.eigh` --- in each case getting slightly different signs for certain eigenvectors. 

Is there a reason that the four methods should or should not agree on the sign convention for the returned eigenvectors?

**Comment by hawkinsp** [comment]
@shoumikdc I don't suppose you could figure out for me which is the first version of scipy that fixed the problem? I think we could simply increase the minimum scipy version constraint for jax, but I can't reproduce the problem myself.

In general if the eigendecomposition is not uniquely defined, any of these functions may return any valid result.

**Comment by shoumikdc** [comment]
@hawkinsp --- actually sorry, I was slightly mistaken! I think the problem is not necessarily with `scipy` but actually with `libopenblas`. 

Check out this [thread](https://github.com/ContinuumIO/anaconda-issues/issues/12855) which also mentions some problems with `openblas` that caused issues with both `numpy.linalg.eigh` and `jnp.linalg.eigh`. See also [here](https://github.com/numpy/numpy/issues/21950) and [here](https://github.com/google/jax/issues/12156).

In particular, the error I showed above is present for `libopenblas` v0.3.20 independent of what `scipy` version I use. However, as suggested in that thread, for the more recent `libopenblas` v0.3.21 onwards, it seems to be fine! 

In particular, I now have a stable version in a different branch that uses `libopenblas` v0.3.23 (the latest) with `scipy=1.10.1` and the error is gone.

**Comment by hawkinsp** [comment]
Awesome! I think there's no action item for us here. Ideally JAX would specify the minimum openblas version, but I don't think there's any way for us to do that since it's not something we can specify as part of `pip` constraints. 

Closing.
